godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,777 @@
+        unsafe { (*ptr).as_ptr() }
+    }
+
+    fn transfer_script_to_godot(script: &Gd<crate::engine::Script>) -> sys::GDExtensionObjectPtr {
+        script.obj_sys()
+    }
+
+    /// # Safety
+    ///
+    /// - `ptr` is expected to point to a list with both a length and capacity of `list_length`.
+    /// - The list pointer `ptr` must have been created with [`transfer_ptr_list_to_godot`].
+    unsafe fn transfer_ptr_list_from_godot<T>(ptr: *const T, list_length: usize) -> Vec<T> {
+        Vec::from_raw_parts(sys::force_mut_ptr(ptr), list_length, list_length)
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid string return pointer.
+    unsafe fn transfer_string_to_godot(string: GString, return_ptr: sys::GDExtensionStringPtr) {
+        string.move_string_ptr(return_ptr);
+    }
+
+    fn transfer_property_state_to_godot(
+        propery_states: Vec<(StringName, Variant)>,
+        property_state_add: sys::GDExtensionScriptInstancePropertyStateAdd,
+        userdata: *mut c_void,
+    ) {
+        let Some(property_state_add) = property_state_add else {
+            return;
+        };
+
+        propery_states.into_iter().for_each(|(name, value)| {
+            let name = ManuallyDrop::new(name);
+            let value = ManuallyDrop::new(value);
+
+            // SAFETY: Godot expects a string name and a variant pointer for each property. After receiving the pointer, the engine is responsible
+            // for managing the memory behind those references.
+            unsafe { property_state_add(name.string_sys(), value.var_sys(), userdata) };
+        });
+    }
+
+    fn transfer_script_lang_to_godot(
+        script_lang: Gd<ScriptLanguage>,
+    ) -> sys::GDExtensionScriptLanguagePtr {
+        script_lang.obj_sys() as sys::GDExtensionScriptLanguagePtr
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide valid pointers to the instance data, property name and value.
+    pub(super) unsafe extern "C" fn set_func<T: ScriptInstance>(
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+        p_name: sys::GDExtensionConstStringNamePtr,
+        p_value: sys::GDExtensionConstVariantPtr,
+    ) -> sys::GDExtensionBool {
+        let name = transfer_string_name_from_godot(p_name);
+        let value = &*Variant::ptr_from_sys(p_value);
+        let ctx = || format!("error when calling {}::set", type_name::<T>());
+
+        let result = handle_panic(ctx, || {
+            let instance = instance_data_as_script_instance::<T>(p_instance);
+
+            borrow_instance_mut(instance).set(name, value)
+        })
+        // unwrapping to a default of false to indicate that the assignment as not handled by the script.
+        .unwrap_or_default();
+
+        transfer_bool_to_godot(result)
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid pointer for the instance data and property name.
+    /// - A valid variant pointer for the return value is expected.
```

Comment:
As written this really isn't a safety invariant. Maybe write this as
```suggestion
    /// - `r_ret` will be a valid owned variant pointer after this call.
```
Same applies for a couple other functions further down.